### ISSUE TITLE
Virtual loss and scaling for GPU-heavy setups

### ISSUE THREAD

**Issue by patrik-ha** [patrik-ha/explainable-minichess#1]
The part of this project that deals with the training pipeline was originally tailored for a mid-range personal workstation. (Meaning a small amount of relatively fast CPU threads, with a low-range GPU) This affected some of the decisions when it came to how the interfacing with the neural nets during training should be done, i.e. giving each thread access to a copy of the neural net on the CPU, essentially only using the GPU for training between epochs. (This is also less of a big deal when the models are small) 

However, for larger models, or larger variants of chess, it would be very useful to better utilise the potential of GPUs, when available. This would mean implementing prediction-batching through virtual loss. This would greatly reduce training time, making it possible to experiment more with different architectures for different goals.

I am currently working on this, as it would be very useful to be able to test such architectures more rapidly.

**Comment by patrik-ha** [comment]
Currently working! Needs to be cleaned up a bit though. It's also not that hard to make the workers keep working all the time. Will get on that

**Comment by patrik-ha** [comment]
Cleaned up a bit as well. I currently prefer the implementation here to the one used in the main branch, but it currently only supports environments that run OpenMPI. It might be easier to understand though, and for users interested in training their own models, it is definitely faster.

**Comment by patrik-ha** [comment]
(do need to test it a bit though)

**Comment by patrik-ha** [comment]
(And it might be a lot faster if workers batched their requests to the GPU-thread too, instead of sending one message per state)
